x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -532,21 +547,19 @@ private void resolveIndicesAndAnalyze(
         PreAnalysisResult result,
         ActionListener<Versioned<LogicalPlan>> logicalPlanListener
     ) {
-        EsqlCCSUtils.initCrossClusterState(indicesExpressionGrouper, verifier.licenseState(), preAnalysis.indexPattern(), executionInfo);
+        EsqlCCSUtils.initCrossClusterState(
+            indicesExpressionGrouper,
+            verifier.licenseState(),
+            preAnalysis.indexes().keySet(),
+            executionInfo
+        );
 
         // The main index pattern dictates on which nodes the query can be executed, so we use the minimum transport version from this field
         // caps request.
         SubscribableListener.<PreAnalysisResult>newForked(
-            l -> preAnalyzeMainIndicesAndRetrieveMinTransportVersion(preAnalysis, executionInfo, result, requestFilter, l)
-        ).andThenApply(r -> {
-            if (r.indices.isValid()
-                && executionInfo.isCrossClusterSearch()
-                && executionInfo.getRunningClusterAliases().findAny().isEmpty()) {
-                LOGGER.debug("No more clusters to search, ending analysis stage");
-                throw new NoClustersToSearchException();
-            }
-            return r;
-        })
+            l -> preAnalyzeMainIndices(preAnalysis.indexes().keySet().iterator(), preAnalysis, executionInfo, result, requestFilter, l)
```

Comment:
Iterating over `entrySet()` could avoid additional lookup later on for the index mode.